content/blog/2019-12-18-halide-autoscheduler/index.md

```diff
@@ -0,0 +1,359 @@
+  loop type is `sequential`, but it could also be `parallel`, `unrolled`,
+  or `vectorized`.
+  For simplicity our implementation only supports `sequential` and `vectorized`.
+
+- **Reorder** - switch loop nodes for the same function.
+
+- **Hoist / lower compute** - change the granularity in which intermediate
+  results are computed.
+
+- **Hoist / lower storage** - change the granularity in which storage for
+  intermediate results is allocated.
+
+- **Inline / deinline** - inline functions into callers (don't store their results
+  in intermediate storage) or deinline function out of callers.
+  Intuitively, inlining functions trades off smaller memory usage for
+  redundant computations, while de-inlining trades off higher memory usage
+  for fewer redundant computations.
+
+
+Below are some diagrams to give intuition to how these scheduler transformers
+work.
+
+![reorder](reorder.png)
+
+![hoist compute](hoist-compute.png)
+
+![lower compute](lower-compute.png)
+
+![hoist storage](lower-compute.png)
+
+![inline](inline.png)
+
+![deinline](deinline.png)
+
+
+### Bounds Inference
+
+Now that we have a representation for schedules and a set of
+schedule transformers, we are close to arriving at a search algorithm
+for finding efficient schedules.
+The last component that we need is a notion of *cost* for schedules.
+In order to provide a cost model for schedules, we need to determine the
+number of iterations performed by loops in the schedule.
+This determines the number of times instructions inside the body of loops
+will be executed, as well as the size of intermediate storage to be
+allocated.
+We determine this by computing the extent in which functions will be computed.
+For the output function, we assume that the extent is given by a call to the
+`realize` function.
+For called functions that are not inlined, the extent is the dimensions of
+the function that will be stored as intermediate results.
+Because storage will be reused depending on the granularity
+with which intermediate results are stored, the extent of called functions
+does not necessarily coincide with the total extent over which the function
+will be computed (e.g., the called function might be computed on
+a per-scanline basis).
+
+For example, consider the simple pipeline below that has one producer (`g`)
+and one consumer (`f`):
+
+`g(x, y) = x * y`
+
+`f(x, y) = g(x, y) + g(x+1, y+1)`
+
+Given that `f` is realized in a 512x512 box and a schedule where `g` is
+computed in total before computing `f`, the extent of `g` is 513x513.
+
+Computing the extent in which functions will be computed is hard in general,
+but since Halide makes the simplifying assumption that all extents are
+rectangular (as opposed to, say, polytopes in the polyhedral model),
+there is a simple method for doing this:
+we only need to check the maximum and minimum points of the caller functions
+and check the arguments to the callee.
```

Comment:
Seems like this also relies on an implicit assumption that the index "propagation" is linear? For example, you might have a function like `f(x) = g(x == 50 ? 1000 : x)` or something, and if you knew the extent for `f` was 0–100, you'd miss the extent of `g` including one far-out point. But maybe this is implicated by the "rectangular" assumption.

Reply:
This function specifically can't be expressed because we assume arguments are drawn from a grammar of "simple" arithmetic expressions --- `+`, `-`, `*`, `/`, variables and constants. Will add text to make this more explicit.